e2e/mock/server.go

```diff
@@ -3,39 +3,78 @@ package mock
 import (
 	"net"
 
-	"github.com/hyperledger/fabric-protos-go/orderer"
-	"github.com/hyperledger/fabric-protos-go/peer"
 	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
 )
 
 type Server struct {
-	GrpcServer *grpc.Server
-	Listener   net.Listener
+	peers   []*Peer
+	orderer *Orderer
 }
 
-func (s *Server) Start() {
+func ConstructServer(peerN int, Credentials credentials.TransportCredentials) (server *Server, err error) {
 	blockC := make(chan struct{}, 1000)
-
-	p := &Peer{
-		BlkSize: 10,
-		TxC:     blockC,
+	server = &Server{}
+	server.peers = make([]*Peer, 0)
+	// for peers
+	for i := 0; i < peerN; i++ {
+		lis, err := net.Listen("tcp", "127.0.0.1:0")
+		if err != nil {
+			return nil, err
+		}
+		grpcServer := grpc.NewServer()
+		if Credentials != nil {
+			grpcServer = grpc.NewServer(grpc.Creds(Credentials))
+		}
+		peer := &Peer{
+			Listener:   lis,
+			GrpcServer: grpcServer,
+			BlkSize:    10,
+			TxC:        blockC,
+		}
+		peer.Prepare()
+		server.peers = append(server.peers, peer)
+	}
+	// for orderer
+	lis, err := net.Listen("tcp", "127.0.0.1:0")
+	if err != nil {
+		return nil, err
 	}
+	grpcServer := grpc.NewServer()
+	if Credentials != nil {
+		grpcServer = grpc.NewServer(grpc.Creds(Credentials))
+	}
+	server.orderer = &Orderer{
+		Listener:   lis,
+		GrpcServer: grpcServer,
+		TxCs:       []chan struct{}{blockC},
+	}
+	server.orderer.Prepare()
+	return server, nil
+}
 
-	o := &Orderer{
-		TxC: blockC,
+func (s *Server) Start() {
+	for _, v := range s.peers {
+		go v.Start()
 	}
+	go s.orderer.Start()
+}
 
-	peer.RegisterEndorserServer(s.GrpcServer, p)
-	peer.RegisterDeliverServer(s.GrpcServer, p)
-	orderer.RegisterAtomicBroadcastServer(s.GrpcServer, o)
+func (s *Server) Stop() {
+	for _, v := range s.peers {
+		v.Stop()
+	}
+	s.orderer.Stop()
+}
 
-	err := s.GrpcServer.Serve(s.Listener)
-	if err != nil {
-		panic(err)
+func (s *Server) GetPeerAdders() (peersAddrs []string) {
```

Comment:
typo: Adders -> Addresses

Reply:
And i think it's sufficiently to have one method `Addresses` to return both peer and orderer addresses. Caller could simply ignore one of return value if not needed